In [7]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from imblearn.over_sampling import RandomOverSampler
from sklearn.metrics import classification_report

cols = ["fLength","fWidth","fSize","fConc","fConc1","fAsym","fM3Long","fM3Trans","fAlpha","fDist","class"]
df=pd.read_csv("magic04.data",names=cols)
df.head()
df['class']=(df['class']=='g').astype(int)
train,valid,test = np.split(df.sample(frac=1),[int(0.6*len(df)),int(0.8*len(df))])

def scale_dataset(dataframe,oversample=False):
    x = dataframe[dataframe.columns[:-1]].values
    y = dataframe[dataframe.columns[-1]].values
    
    scaler = StandardScaler()
    x= scaler.fit_transform(x)
    ros =RandomOverSampler()
    x,y = ros.fit_resample(x,y)
    
    data = np.hstack((x,np.reshape(y,(-1,1))))
    
    return data,x,y

train,x_train,y_train  = scale_dataset(train,oversample=True)
valid,x_valid,y_valid  = scale_dataset(valid,oversample=False)
test,x_test,y_test  = scale_dataset(test,oversample=False)

c:\Users\RBTG\AppData\Local\Programs\Python\Python312\Lib\site-packages\numpy\core\fromnumeric.py:59: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.
  return bound(*args, **kwds)


In [14]:
from sklearn.neighbors import KNeighborsClassifier
knn_model = KNeighborsClassifier(n_neighbors=4)
knn_model.fit(x_train,y_train)
y_predict = knn_model.predict(x_test)
print("Predict values ",y_predict)
print("Test values ",y_test)
print(classification_report(y_test,y_predict))

Predict values  [0 1 1 ... 1 1 0]
Test values  [0 1 1 ... 0 0 0]
              precision    recall  f1-score   support

           0       0.80      0.76      0.78      2448
           1       0.77      0.81      0.79      2448

    accuracy                           0.79      4896
   macro avg       0.79      0.79      0.79      4896
weighted avg       0.79      0.79      0.79      4896



##### In the context of classification, particularly binary classification, terms like "true positives," "false positives," "true negatives," and "false negatives" are used to describe the outcomes of the classification process. Let's define these terms and clarify what is meant by "true positive prediction" and "total predicted positives."

### True Positive (TP)

#### True Positive Prediction:
This occurs when the model correctly predicts a positive class. In other words, the model's prediction and the actual class are both positive.
#### Example: 
If the model is predicting whether an email is spam (positive class), a true positive is when the model correctly identifies a spam email as spam.

### False Positive (FP)


#### False Positive: 
This occurs when the model incorrectly predicts a positive class, while the actual class is negative. It's also known as a "Type I error."
#### Example: 
If the model incorrectly labels a legitimate email as spam, that's a false positive.

### True Negative (TN)
#### True Negative: 
This occurs when the model correctly predicts a negative class. The model's prediction and the actual class are both negative.
#### Example: 
Correctly identifying a legitimate email as not spam.
#### False Negative (FN)
#### False Negative: 
This occurs when the model incorrectly predicts a negative class, while the actual class is positive. It's also known as a "Type II error."
#### Example: 
If the model fails to identify a spam email and labels it as not spam, that's a false negative.
#### Precision
Precision is calculated as the ratio of true positives to the total number of instances that were predicted as positive. It measures the accuracy of the positive predictions.

Precision
=
True Positives (TP)
True Positives (TP)
+
False Positives (FP)
Precision= 
True Positives (TP)+False Positives (FP)
True Positives (TP)
​
 

#### True Positives (TP): 
The number of correct positive predictions.
#### Total Predicted Positives: 
The sum of true positives and false positives, representing all instances the model predicted as positive.
#### Example
Let's say we have a model that predicts whether a tumor is malignant (positive) or benign (negative). If the model predicts 100 cases as malignant, and out of those 100, 80 are actually malignant (true positives) and 20 are benign (false positives), then:

True Positives (TP) = 80
Total Predicted Positives = TP + FP = 80 + 20 = 100
Precision in this case would be:

Precision
=
80/
100
=
0.80
Precision= 
100/
80
​
 =0.80

This means that 80% of the time, when the model predicts a tumor is malignant, it is correct.